src/Staged/GHC/Generics/GHCish.hs

```diff
@@ -0,0 +1,149 @@
+              (normalB $ unTypeCode $ k (M2 (R2 (M2
+                (M2 (K2 (unsafeCodeCoerce $ varE v1)) :**: M2 (K2 (unsafeCodeCoerce $ varE v2)))))))
+              []
+    ]
+-}
+
+-- | Use GHC generics to implement the 'from' method.
+ghcGenericFrom :: (Rep a ~ Translate (G.Rep a), G.Generic a, GGeneric (Rep a), Quote q)
+  => Code q a -> (Rep a (Code q) x -> Code q r) -> Code q r
+ghcGenericFrom c k = unsafeCodeCoerce $ caseE (unTypeCode c) $ gmatches k
+
+class GGeneric (f :: (Type -> Type) -> Type -> Type) where
+  gto :: Quote q => f (Code q) x -> q Exp
+
+  -- Build the list of case branches.
+  gmatches :: Quote q => (f (Code q) x -> Code q r) -> [q Match]
+
+-- | Make a "namer" function holding on to a package and module name to
+-- eventually make a Name out of the bare constructor name
+mkNamer :: forall (d :: Meta). Datatype d => String -> Name
+mkNamer = mkNameG_d (packageName prox) (moduleName prox)
+  where
+    prox :: Prox d z w
+    prox = Prox
+
+instance (Datatype d, GGenericCon f) => GGeneric (D2 d f) where
+  gto (M2 fqp) = gtoCon (mkNamer @d) fqp
+  gmatches k = gmatchesCon (mkNamer @d) (k . M2) []
+
+class GGenericCon (f :: (Type -> Type) -> Type -> Type) where
+  gtoCon :: Quote q => (String -> Name) -> f (Code q) x -> q Exp
+  -- We take a "namer" function that holds on to the package and module names
+  -- for us so we can apply the constructor name and get its 'Name'.
+  gmatchesCon :: Quote q => (String -> Name) -> (f (Code q) x -> Code q r) -> [q Match] -> [q Match]
+
+instance GGenericCon V2 where
+  gtoCon _ x = case x of
+  gmatchesCon _ _ [] = []
+  gmatchesCon _ _ _ = error "V2 shouldn't lead to more constructors!"
+
+instance (GGenericCon f, GGenericCon g) => GGenericCon (f :++: g) where
+  gtoCon namer (L2 x) = gtoCon namer x
+  gtoCon namer (R2 x) = gtoCon namer x
+
+  gmatchesCon namer k = gmatchesCon @f namer (k . L2) . gmatchesCon @g namer (k . R2)
+
+type GNumFields :: forall {k}. ((Type -> Type) -> k -> Type) -> Constraint
+class GNumFields (f :: (Type -> Type) -> k -> Type) where
+  gnumFields :: Int
+instance (GNumFields f, GNumFields g) => GNumFields (f :**: g) where
+  gnumFields = gnumFields @f + gnumFields @g
+instance GNumFields U2 where
+  gnumFields = 0
+instance GNumFields (S2 c f) where
+  gnumFields = 1
+
+instance (Constructor c, GGenericFields f, GNumFields f) => GGenericCon (C2 c f) where
+  gtoCon namer m@(M2 fqp) = gtoFields (conE conN) fqp
+    where
+      conN :: Name
+      conN = namer (conName (toProx m))
+  gmatchesCon namer k rest = [do
+    let name_bases = ["v" ++ show n | n <- [1..gnumFields @f]]
+    names <- traverse newName name_bases
+    match (conP conN (map varP names)) (normalB . unTypeCode . k . M2 $ fst (grebuild names)) []
+    ] ++ rest
+    where
+      conN :: Name
+      conN = namer (conName (Prox @_ @c))
+
+class GGenericFields f where
+  gtoFields :: Quote q => q Exp -> f (Code q) x -> q Exp
+  grebuild :: Quote q => [Name] -> (f (Code q) x, [Name])
```

Comment:
I think this can be written with

```haskell
grebuild :: Quote q => f Identity Name -> f (Code q)
```

type (or you could use GHC.Generics functors for name container.

Then `GNumFields` can be changed to generate the initial `f Identity Name`. 

That should make the code look less unsafe: no ned for `error "oopsy"` as name container and build result cardinality would match.

Reply:
I tried to figure out a way to do this nicely, but failed. I don't think `Identity` will get enough variables, because that only goes one product deep. I too suspect something better is possible; I just don't know what is supposed to look like. Maybe mumble mumble free applicatives mumble.

Reply:
ok. i'll try myself later.

Reply:
Er... No, sorry, that's not what I meant exactly...

Reply:
No, no luck yet, and I have to do other things now. Give it a try if you like.

Reply:
https://github.com/phadej/staged-gg/commit/fd25848ab33610f1fcbe5db8a574b06b93f3aa6f

Reply:
`gtoFields` could also be simplified to do `f (Code q) x -> `[q Exp]`, if there were a foldMap like function to do fold over `f`s. Then the `gtoCon` could just `appsE`.

I don't write one and rather cheat in `makeNames`, but it could be used there too.

All very general.

Reply:
Very clever. I forgot how `q` fits in! My brain was stuck in `GHC.Generics`.